distributed/worker_memory.py

```diff
@@ -0,0 +1,391 @@
+                # dependency. We could have alternatively used a weakref.
+                # FIXME annotations: https://github.com/tornadoweb/tornado/issues/3117
+                partial(self.memory_monitor, worker),  # type: ignore
+                self.memory_monitor_interval * 1000,
+            )
+            worker.periodic_callbacks["memory_monitor"] = pc
+
+        self._throttled_gc = ThrottledGC(logger=logger)
+
+    async def memory_monitor(self, worker: Worker) -> None:
+        """Track this process's memory usage and act accordingly.
+        If process memory rises above the spill threshold (70%), start dumping data to
+        disk until it goes below the target threshold (60%).
+        If process memory rises above the pause threshold (80%), stop execution of new
+        tasks.
+        """
+        if self._memory_monitoring:
+            return
+        self._memory_monitoring = True
+        # Don't use psutil directly; instead read from the same API that is used to send
+        # info to the Scheduler (e.g. for the benefit of Active Memory Manager) and
+        # which can be easily mocked in unit tests.
+        memory = worker.monitor.get_process_memory()
+        self._maybe_pause_or_unpause(worker, memory)
+        await self._maybe_spill(worker, memory)
+        self._memory_monitoring = False
+
+    def _maybe_pause_or_unpause(self, worker: Worker, memory: int) -> None:
+        if self.memory_pause_fraction is False:
+            return
+
+        assert self.memory_limit
+        frac = memory / self.memory_limit
+        # Pause worker threads if above 80% memory use
+        if frac > self.memory_pause_fraction:
+            # Try to free some memory while in paused state
+            self._throttled_gc.collect()
+            if worker.status == Status.running:
+                logger.warning(
+                    "Worker is at %d%% memory usage. Pausing worker.  "
+                    "Process memory: %s -- Worker memory limit: %s",
+                    int(frac * 100),
+                    format_bytes(memory),
+                    format_bytes(self.memory_limit)
+                    if self.memory_limit is not None
+                    else "None",
+                )
+                worker.status = Status.paused
+        elif worker.status == Status.paused:
+            logger.warning(
+                "Worker is at %d%% memory usage. Resuming worker. "
+                "Process memory: %s -- Worker memory limit: %s",
+                int(frac * 100),
+                format_bytes(memory),
+                format_bytes(self.memory_limit)
+                if self.memory_limit is not None
+                else "None",
+            )
+            worker.status = Status.running
+
+    async def _maybe_spill(self, worker: Worker, memory: int) -> None:
+        if self.memory_spill_fraction is False:
+            return
+        if not isinstance(self.data, SpillBuffer):
+            return
```

Comment:
Should this check be removed? Dask CUDA  doesn't use a `SpillBuffer`, as noted [here](https://github.com/dask/distributed/pull/5904/files#r820909243).

Reply:
Alternately, Dask CUDA should inherit from `SpillBuffer` - in which case, perhaps it'd be a good idea to define what interface a `SpillBuffer` should implement? Either as part of the docs, or codified as an ABC?

Reply:
@shwina to answer the question I would need to understand what Dask CUDA is trying to do and what it was doing before.

Is it supposed to react to the ``distributed.worker.memory.spill`` threshold?

If so, how does it do it? Does it replace the ``memory_monitor`` method, or is it duck-type compatible with SpillBuffer and offer a ``data.fast.evict()`` method? I couldn't find either. If neither is the case, does it mean that, even before #5543 and #5736, ``Worker.memory_monitor`` was crashing?

BTW I noticed now that it's not wrapped with ``log_errors``, so the crash would be completely invisible. I'm adding it now.

Reply:
Can you point us to where the Dask CUDA data mapping is defined?

Reply:
Dast CUDA implements two different data mappings to handle GPU->CPU spilling:
- https://github.com/rapidsai/dask-cuda/blob/branch-22.04/dask_cuda/device_host_file.py#L145
- https://github.com/rapidsai/dask-cuda/blob/branch-22.04/dask_cuda/proxify_host_file.py#L433

They implement [`data.fast.evict()`](https://github.com/rapidsai/dask-cuda/blob/branch-22.04/dask_cuda/proxify_host_file.py#L624).

Reply:
I see that ProxifyHostFile has been modified to support both before and after #5543 (2022.02.1).
DeviceHostFile won't work with >=2022.02.1 as it misses an evict() method.

See my latest commit ([TODO](https://github.com/dask/distributed/pull/5904/commits/a8ac78df960a752f4bbd4075538147c147bd6a2f)) where I formalise this interface (missing high level docs and unit tests). Could you confirm that ProxifyHostFile works again now?

Reply:
> DeviceHostFile won't work with >=2022.02.1 as it misses an evict() method.

Implemented now in https://github.com/shwina/dask-cuda/pull/2